app/gui/src/project-view/components/GraphEditor/detaching.ts

```diff
@@ -0,0 +1,64 @@
+import type { GraphDb, NodeId } from '$/providers/openedProjects/graph/graphDatabase'
+import { set } from 'lib0'
+import { isIdentifier, type AstId, type Identifier } from 'ydoc-shared/ast'
+
+const none: unique symbol = Symbol()
+type None = typeof none
+
+/**
+ * Return all changes to port values for "connecting around" operation of given selected nodes.
+ *
+ * The connections going out of the `selected` group will be reconnected to their self-port
+ * input connection. If any connection could not be subsituted this way (e.g. no connection to self-port),
+ * an empty array will be returned.
+ */
+export function analyzeConnectAround(selected: Set<NodeId>, graphDb: GraphDb) {
+  const knownMainSourceIdentifier = new Map<NodeId, Identifier | None>()
+  const result: { port: AstId; ident: Identifier }[] = []
+
+  const findMainSourceOf = (selectedNode: NodeId) => {
+    const findResult = (): Identifier | None => {
+      const alreadyKnown = knownMainSourceIdentifier.get(selectedNode)
+      if (alreadyKnown) return alreadyKnown
+      const selfPort = graphDb.nodeIdToNode.get(selectedNode)?.primaryApplication.selfArgument
+      if (!selfPort) return none
+      const mainSourceId = set.first(graphDb.connections.reverseLookup(selfPort))
+      if (!mainSourceId) return none
+      const mainSourceNode = graphDb.getPatternExpressionNodeId(mainSourceId)
+      if (mainSourceNode && selected.has(mainSourceNode)) {
+        return findMainSourceOf(mainSourceNode)
+      } else {
+        const ident = graphDb.getOutputPortIdentifier(mainSourceId)
+        return ident != null && isIdentifier(ident) ? ident : none
+      }
+    }
+    const result = findResult()
+    knownMainSourceIdentifier.set(selectedNode, result)
+    return result
+  }
```

Comment:
@kazcw I considered your comment in https://github.com/enso-org/enso/pull/14325#discussion_r2568839061 but sorting here would require bringing method function too (with guarantee that this is _same_ method as graphDb), so I prefer to keep this, more self-contained solution.